### DISCUSSION TITLE
Removal of FastCollections

### DISCUSSION THREAD

**Discussion by IXLLEGACYIXL** [stride3d/stride#2298] [Ideas]
currently we have fast list , fast dictionary , fast list struct which is overcomplicating stuff as it can be replaced with normal collections anyway

1. are these collections actually fast?
2. are these even needed?
3. does anyone know when to use them anyway? havent seen them being used in a recent PR

also they are a source of potential bugs

**Comment by Jklawreszuk** [comment]
1. Would be cool to analyze  whats the difference between list vs fastlist, dictionary vs fastdict and perform simple benchmark tests
2. Yes, they are.  These collections are strongly related to the engine. The question is whether they are actually "fast", because we haven't visited quite long time.
3. Practical use : TrackingCollection.cs or Entity.cs

**Reply by IXLLEGACYIXL** [reply]
ye tracking collection is a special case which is actually required

fast list doesnt seem to have anything special compared to list
so probably excluding fast tracking collection for now from the idea is better

**Reply by Jklawreszuk** [reply]
@IXLLEGACYIXL I completely agree,  It looks very similar to the .NET List , but it would be still worth checking using Benchmark.NET for difference in performance 👨‍🔬

**Reply by Jklawreszuk** [reply]
That applies to other "fast" structures ofc , just to be sure

**Reply by Doprez** [reply]
I did a naive test with Benchmark dotnet with just FastList and a regular List of ints. And the base .NET lists are consistantly faster, I am very new to using these benchmarking tools though so it would be nice to have a second opinion.

![image](https://github.com/stride3d/stride/assets/73259914/27b784d4-805a-41f2-8173-fd9177979918)

```
[RPlotExporter]
public class BenchMarkTest
{
	public FastList<int> FastInts;

	public List<int> RegularInts;

	[GlobalSetup]
	public void Setup()
	{
		FastInts = [];
		for (int i = 0; i < 1000000; i++)
		{
			FastInts.Add(i);
		}
		RegularInts = [];
		for (int i = 0; i < 1000000; i++)
		{
			RegularInts.Add(i);
		}
	}

	[Benchmark]
	public void TestUpdatingFastList_1000()
	{
		for (int i = 0; i < 1000; i++)
		{
			FastInts[i] = i;
		}
	}

	[Benchmark]
	public void TestUpdatingRegularList_1000()
	{
		for (int i = 0; i < 1000; i++)
		{
			RegularInts[i] = i;
		}
	}

	[Benchmark]
	public void TestUpdatingFastList_1000000()
	{
		for (int i = 0; i < 1000000; i++)
		{
			FastInts[i] = i;
		}
	}

	[Benchmark]
	public void TestUpdatingRegularList1000000()
	{
		for (int i = 0; i < 1000000; i++)
		{
			RegularInts[i] = i;
		}
	}

	[Benchmark]
	public void ReinitialiseFastList_1000()
	{
		FastInts = [];
		for (int i = 0; i < 1000; i++)
		{
			FastInts.Add(i);
		}
	}

	[Benchmark]
	public void ReinitialiseRegularList_1000()
	{
		RegularInts = [];
		for (int i = 0; i < 1000; i++)
		{
			RegularInts.Add(i);
		}
	}
}
```

**Reply by Eideren** [reply]
Looks pretty good, the last thing one would have to test is whether iterating over `var items = FastInts.Items;` is faster than iterating over a list.

**Reply by Doprez** [reply]
ahhhh ok, good call.
![image](https://github.com/stride3d/stride/assets/73259914/ebbea5d8-1dd4-4f75-a6b6-ecc548a33bb2)

Added the below tests
```csharp
[Benchmark]
public void TestUnderlyingArrayFastList_1000()
	{
for (int i = 0; i < 1000; i++)
	{
		FastInts.Items[i] = i;
	}
}

[Benchmark]
public void TestUnderlyingArrayFastList_1000000()
{
	for (int i = 0; i < 1000000; i++)
	{
		FastInts.Items[i] = i;
	}
}
```

So it looks like there is a decent difference between the two when you actually access the underlying array.

**Reply by Eideren** [reply]
Right, but now we have https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan which does the same thing for `List<T>`, so I think we can deprecate FastList and replace calls to `FastList<T>.Items` with `AsSpan` above.
We can close this discussion and open an issue with those findings, and/or create a PR if anyone wants to take over.

**Reply by Doprez** [reply]
I guess the work would to be first add the deprecated attribute to the FastList with a recommendation and then start working through replacing them throughout the code eh?

**Reply by Doprez** [reply]
![image](https://github.com/stride3d/stride/assets/73259914/cbc58c9c-406b-4465-a928-d65fe006928c)

holy cow yeah, this is beautiful.

**Comment by Basewq** [comment]
I think there was only one real reason for them which was implementing `struct Enumerator`.
The standard collections have class enumerators, so `foreach` always allocates objects. There's been a few posts in the past with XNA/Monogame/Unity how you must avoid the standard collections because of this (or only use `for (int i = 0...)`).

**Reply by Eideren** [reply]
Looks like they've always been structs from Framework 2.0 onwards: https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.enumerator?view=netframework-2.0

**Reply by Basewq** [reply]
Hmm, I must be forgetting some of the details then...
I think the advice might have been when interacting linq code (which pushed the class up to IList or whatever), though I think that has been changed to use the struct versions when it detects the standard collections nowadays.

I do recall that Stride's `FastList` exposed the underlying array because it pushed it out to some graphics buffer or something, but that was the only 'fast' thing about it.

**Comment by Eideren** [comment]
FastList [provides its backing array for read and write operations](https://github.com/stride3d/stride/blob/master/sources/core/Stride.Core/Collections/FastList.cs#L30) for example, which at the time was not something the BCL provided, now we have the collection marshal namespace providing unsafe access to it.
At the time those types were created collections were pretty awful performance wise. Dotnet has really been focusing on optimizing them since then, I doubt that ours would be faster than dotnet's but we'll have to run benchmarks to ensure that's the case before deprecating them.

**Reply by Kryptos-FR** [reply]
If I remember correctly, it was the main reason for creating those alternative implementations. At the time, if we had used the standard collection we would have had to `ToArray()` them forcing an expensive copy/allocation.

If it can be replaced with officially supported code, then we should indeed change that. We must be careful to properly list all usages to not introduce performance regressions.
